File: src/structured.ts
import { createOpenRouter } from '@openrouter/ai-sdk-provider';
import { generateObject } from 'ai';
import { z } from 'zod';
import * as dotenv from 'dotenv';

dotenv.config();

const openrouter = createOpenRouter({
  apiKey: process.env.OPENROUTER_API_KEY || 'YOUR_OPENROUTER_API_KEY',
});

async function example1_BasicStructuredData() {
  console.log('\n=== Example 1: Basic Structured Data ===\n');

  const { object } = await generateObject({
    model: openrouter.chat('minimax/minimax-m2'),
    schema: z.object({
      recipe: z.object({
        name: z.string(),
        ingredients: z.array(
          z.object({
            name: z.string(),
            amount: z.string(),
          }),
        ),
        steps: z.array(z.string()),
      }),
    }),
    prompt: 'Generate a recipe for chocolate chip cookies. Output as JSON.',
  });

  console.log('Recipe:', JSON.stringify(object, null, 2));
}

async function example2_ComplexStructuredData() {
  console.log('\n=== Example 2: Complex Structured Data ===\n');

  const { object } = await generateObject({
    model: openrouter.chat('minimax/minimax-m2'),
    schema: z.object({
      characters: z.array(
        z.object({
          name: z.string(),
          class: z.enum(['warrior', 'mage', 'archer', 'rogue']),
          level: z.number(),
          stats: z.object({
            strength: z.number(),
            intelligence: z.number(),
            dexterity: z.number(),
          }),
          equipment: z.array(z.string()),
        }),
      ),
    }),
    prompt: 'Generate 3 RPG characters with different classes. Output as JSON.',
  });

  console.log('Characters:', JSON.stringify(object, null, 2));
}

async function example3_NotificationData() {
  console.log('\n=== Example 3: Notification Data ===\n');

  const { object } = await generateObject({
    model: openrouter.chat('minimax/minimax-m2'),
    schema: z.object({
      notification: z.object({
        title: z.string(),
        body: z.string(),
        actions: z.array(
          z.object({
            label: z.string(),
            action: z.string(),
          }),
        ),
      }),
    }),
    prompt: 'Generate a notification for a user receiving a new message. Output as JSON.',
  });

  console.log('Notification:', JSON.stringify(object, null, 2));
}

async function example4_ArrayData() {
  console.log('\n=== Example 4: Array of Items ===\n');

  const { object } = await generateObject({
    model: openrouter.chat('minimax/minimax-m2'),
    schema: z.object({
      tasks: z.array(
        z.object({
          id: z.number(),
          title: z.string(),
          priority: z.enum(['low', 'medium', 'high']),
          estimatedHours: z.number(),
          tags: z.array(z.string()),
        }),
      ),
    }),
    prompt: 'Generate 5 software development tasks for building a todo app. Output as JSON.',
  });

  console.log('Tasks:', JSON.stringify(object, null, 2));
}

async function main() {
  try {
    await example1_BasicStructuredData();
    await example2_ComplexStructuredData();
    await example3_NotificationData();
    await example4_ArrayData();
  } catch (error) {
    console.error('Error:', error instanceof Error ? error.message : String(error));
  }
}

main();

File: python/examples/structured.py
"""Structured output (JSON mode) example with MiniMax M2.

Equivalent to TypeScript src/structured.ts
Note: Free tier uses JSON mode instead of formal structured outputs
"""

import sys
import os
import json

# Add parent directory to path to import utils
sys.path.insert(0, os.path.abspath(os.path.join(os.path.dirname(__file__), '..')))

from utils.config import Config
from pydantic import BaseModel
from typing import List, Literal


# Define Pydantic models
class Ingredient(BaseModel):
    name: str
    amount: str


class Recipe(BaseModel):
    name: str
    ingredients: List[Ingredient]
    steps: List[str]


class RecipeResponse(BaseModel):
    recipe: Recipe


def example1_basic_structured():
    """Example 1: Basic structured data - Recipe."""
    print("\n=== Example 1: Basic Structured Data (Recipe) ===\n")

    client = Config.get_client()

    # Define the expected schema
    schema = RecipeResponse.model_json_schema()

    response = client.chat.completions.create(
        model=Config.MODEL_FREE,
        messages=[
            {
                "role": "system",
                "content": f"Output valid JSON matching this schema: {json.dumps(schema)}"
            },
            {
                "role": "user",
                "content": "Generate a recipe for chocolate chip cookies. Output as JSON."
            }
        ],
        response_format={"type": "json_object"}
    )

    # Parse and validate with Pydantic
    result_dict = json.loads(response.choices[0].message.content)
    result = RecipeResponse(**result_dict)

    print("Recipe:", json.dumps(result.model_dump(), indent=2))


class Task(BaseModel):
    id: int
    title: str
    priority: Literal["low", "medium", "high"]
    estimatedHours: float
    tags: List[str]


class TaskList(BaseModel):
    tasks: List[Task]


def example2_array_data():
    """Example 2: Array of structured items."""
    print("\n=== Example 2: Array of Tasks ===\n")

    client = Config.get_client()

    schema = TaskList.model_json_schema()

    response = client.chat.completions.create(
        model=Config.MODEL_FREE,
        messages=[
            {
                "role": "system",
                "content": f"Output valid JSON matching this schema: {json.dumps(schema)}"
            },
            {
                "role": "user",
                "content": "Generate 5 software development tasks for building a todo app. Output as JSON."
            }
        ],
        response_format={"type": "json_object"}
    )

    result_dict = json.loads(response.choices[0].message.content)
    result = TaskList(**result_dict)

    print("Tasks:", json.dumps(result.model_dump(), indent=2))


def main():
    """Run all structured output examples."""
    print("Testing MiniMax M2 Free Model - Structured Output (JSON Mode)\n")
    print("=" * 60)

    try:
        example1_basic_structured()
        example2_array_data()

        print("\n" + "=" * 60)
        print("All examples completed!")
        print("=" * 60)

    except Exception as error:
        print(f"Error: {error}")


if __name__ == "__main__":
    main()
